### ISSUE TITLE
Is it possible to use asyncio/threading with web3 contract calls?

### ISSUE THREAD

**Issue by gohuygo** [ethereum/web3.py#1764]
* Version: 5.12.1
* Python: 3.7.4
* OS: osx
* `pip freeze` output

### What was wrong?

I have some piece of code that I want to optimize for performance. The bottleneck currently happens whenever I make calls with web3py. (e.g. web3.eth.blockNumber or my smart contract functions).

Example:

```
class Strategy:
    def __init__(
        self,
        ethereum: Ethereum,
    ) -> None:
        self.ethereum = ethereum

    def process(self):
        while True:
            all_paths = self._load_paths()
            asyncio.run(self._queue_tasks(all_paths))

    async def _queue_tasks(self, all_paths):
        async with aiohttp.ClientSession() as session:
            coroutines = [self._task(index, chunk) for index, chunk in enumerate(self._chunk(all_paths, 10))]
            pending_tasks = set(
                [asyncio.create_task(coro) for coro in coroutines]
            )
        while pending_tasks:
            done_tasks, pending_tasks = await asyncio.wait(pending_tasks,  return_when=asyncio.ALL_COMPLETED)

    async def _task(self, thread, paths):

        counter = 1
        current_block = self.ethereum.w3.eth.blockNumber

        while counter < 200:
            print("COUNTER", thread, counter)
            latest_block = wait_new_block(self.ethereum, current_block)
            current_block = latest_block

            gas_price = self.ethereum.w3.eth.gasPrice * int(gas_price * 1.5)

            self.my_contract.call_contract_function(
                paths, latest_block, gas_price
            )

            counter += 1
```

Each iteration of `_queue_tasks` currently takes a few seconds to run and is IO-bound. I can run some of these in parallel while waiting on blockchain calls to improve my service's performance.

I would like to do something like `await self.ethereum.w3.eth.blockNumber` but it looks like the async API work slowed down and died a few years ago - https://github.com/ethereum/web3.py/milestone/6

Are there any workarounds to await calls to Ethereum via web3.py? Appreciate any tips.

### How can it be fixed?

Fill this section in if you know how this could or should be fixed.

**Comment by kclowes** [comment]
@gohuygo there isn't really any native async support in web3 yet, but you can wrap some functions. Here is an example: https://web3py.readthedocs.io/en/stable/filters.html?#single-threaded-concurrency-with-async-and-await. Closing, but feel free to reopen either this or another issue if you have more questions!

**Comment by qiujiangkun** [comment]
FYI, I call jsonrpc with my async http client.
```python
from tornado import httpclient

http_client = httpclient.AsyncHTTPClient()

async def jsonrpc_send_msg(method: str, *params) -> Result[Any, Web3Error]:
    try:
        response = await http_client.fetch(
            "https://proxy.roninchain.com/free-gas-rpc",
            method="POST",
            headers=headers,
            body=json.dumps(
                {"id": 0, "jsonrpc": "2.0", "method": method, "params": params}
            ),
        )
        result = json.loads(response.body)
        logger.debug("jsonrpc_send_msg: %s %s %s", method, params, result)
        if result.get("error"):
            return Err(
                JsonRpcError(result["error"]["code"], result["error"]["message"])
            )
        return Ok(result["result"])
    except httpclient.HTTPError as err:
        error_msg = err.response and err.response.body.decode() or "<empty response>"
        if error_msg == "free gas request times is exceeded":
            return Err(RoninNoFreeGasError())
        return Err(NetworkError(err, method + " " + error_msg + " " + str(params), err))

@beartype.beartype
async def check_transaction_receipt(hash: str) -> Result[str, Web3Error]:
    receipt = await jsonrpc_send_msg("eth_getTransactionReceipt", hash)
    if isinstance(receipt, Err):
        return receipt
    value = receipt.ok()
    if value is None:
        return Err(TransactionWaitMiningError(hash))
    if value.get("status") == "0x0":
        return Err(TransactionFailed(hash, value))
    return receipt
@beartype.beartype
async def send_slp(
    frm_orig: str, to_orig: str, value: int, priv: str
) -> Result[str, Web3Error]:
    frm = w3.toChecksumAddress(frm_orig.replace("ronin:", "0x"))
    to = w3.toChecksumAddress(to_orig.replace("ronin:", "0x"))
    nonce = await jsonrpc_send_msg("eth_getTransactionCount", frm, "pending")
    if isinstance(nonce, Err):
        return nonce
    else:
        nonce = nonce.ok()
    transaction = contract.functions.transfer(to, value).buildTransaction(
        {
            "chainId": 2020,
            "from": frm,
            "gas": 38000,
            "gasPrice": 0,
            "nonce": nonce,
        }
    )

    signed_txn = w3.eth.account.sign_transaction(transaction, priv)
    resp = await jsonrpc_send_msg(
        "eth_sendRawTransaction", signed_txn.rawTransaction.hex()
    )
    if resp.err():
        return resp
    return Ok(signed_txn.hash.hex())

```
